src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs

```diff
@@ -116,25 +139,30 @@ public class DiagnosticPipeline : IDisposable
                     filterTransform = new FilterAction(
                         sink.Filters,
                         this.cancellationTokenSource.Token,
-                        this.pipelineConfiguration.MaxEventBatchSize,
+                        MaxNumberOfBatchesInProgress,
                         this.pipelineConfiguration.MaxConcurrency,
-                        healthReporter);
+                        healthReporter,
+                        this.OnEventsFilteredOut);
                     var filterBlock = filterTransform.GetFilterBlock();
-                    this.pipelineDisposables.Add(sinkSource.LinkTo(filterBlock, propagateCompletion));
+                    this.pipelineLinkDisposables.Add(sinkSource.LinkTo(filterBlock, propagateCompletion));
                     outputSource = filterBlock;
                 }
 
                 OutputAction outputAction = new OutputAction(
                     sink.Output,
                     this.cancellationTokenSource.Token,
-                    this.pipelineConfiguration.MaxEventBatchSize,
+                    MaxNumberOfBatchesInProgress,
                     this.pipelineConfiguration.MaxConcurrency,
-                    healthReporter);
-                ActionBlock<EventData[]> outputBlock = outputAction.GetOutputBlock();
-                this.pipelineDisposables.Add(outputSource.LinkTo(outputBlock, propagateCompletion));
+                    healthReporter,
+                    (eventsSentCount) => Interlocked.Add(ref this.eventsInProgress, -eventsSentCount));
```

Comment:
Interlocked method do not use any locks.

I run the benchmark and there is no discernible effect on performance.